nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs

```diff
@@ -52,3 +74,172 @@ impl FunctionConstructor {
+    kind: DynamicFunctionKind,
+    parameter_args: &[Value],
+    body_arg: Value,
+) -> JsResult<ECMAScriptFunction> {
+    // 11. Perform ? HostEnsureCanCompileStrings(currentRealm, parameterStrings, bodyString, false).
+    agent
+        .host_hooks
+        .host_ensure_can_compile_strings(agent.current_realm_mut())?;
+
+    let parameters = {
+        let mut parameters = std::string::String::new();
+        for (i, parameter) in parameter_args.iter().enumerate() {
+            if i != 0 {
+                parameters.push(',');
+            }
+            parameters.push_str(parameter.to_string(agent)?.as_str(agent));
+        }
+        parameters
+    };
+    let source_string = format!(
+        "{} anonymous({}\n) {{\n{}\n}}",
+        kind.prefix(),
+        parameters,
+        body_arg.to_string(agent)?.as_str(agent)
+    );
+    let source_string = String::from_string(agent, source_string);
+
+    // The spec says to parse the parameters and the function body separately to
+    // avoid code injection, but oxc doesn't have a public API to do that.
+    // Instead, we parse the source string as a script, and throw unless it has
+    // exactly one statement which is a function declaration of the right kind.
+    let (function, source_code) = {
+        let mut function = None;
+        let mut source_code = None;
+
+        // SAFETY: The safety requirements are that the SourceCode cannot be
+        // GC'd before the program is dropped. If this function returns
+        // successfully, then the program's AST and the SourceCode will both be
+        // kept alive in the returned function object.
+        let parsed_result =
+            unsafe { SourceCode::parse_source(agent, source_string, Default::default()) };
+
+        if let Ok((program, sc)) = parsed_result {
+            source_code = Some(sc);
+            if program.hashbang.is_none()
+                && program.directives.is_empty()
+                && program.body.len() == 1
+            {
+                if let oxc_ast::ast::Statement::FunctionDeclaration(funct) = &program.body[0] {
+                    if kind.function_matches_kind(funct) {
+                        // SAFETY: the Function is inside a oxc_allocator::Box, which will remain
+                        // alive as long as `source_code` is kept alive. Similarly, the inner
+                        // lifetime of Function is also kept alive by `source_code`.`
+                        function = Some(unsafe {
+                            std::mem::transmute::<
+                                &oxc_ast::ast::Function,
+                                &'static oxc_ast::ast::Function,
+                            >(funct)
+                        });
+                    }
+                }
+            }
+        }
+
+        if let Some(function) = function {
+            (function, source_code.unwrap())
+        } else {
+            return Err(agent.throw_exception_with_static_message(
```

Comment:
**nitpick:** We haven't really been doing this but we should and could: We know that SourceCode cannot be referred to by any other anywhere, and that it is the last item in the source_codes Vec. We can thus pop it off the Vec before returning, so as to help GC along.

Reply:
Done.